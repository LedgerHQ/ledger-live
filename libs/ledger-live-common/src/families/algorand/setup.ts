// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { createBridges } from "@ledgerhq/coin-algorand/bridge/js";
import makeCliTools from "@ledgerhq/coin-algorand/cli-transaction";
import algorandResolver from "@ledgerhq/coin-algorand/hw-getAddress";
import { Transaction } from "@ledgerhq/coin-algorand/types";
import Algorand from "@ledgerhq/hw-app-algorand";
import Transport from "@ledgerhq/hw-transport";
import network from "@ledgerhq/live-network/network";
import type { Bridge } from "@ledgerhq/types-live";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import type { Resolver } from "../../hw/getAddress/types";

const createSigner: CreateSigner<Algorand> = (transport: Transport) => {
  return new Algorand(transport);
};

const bridge: Bridge<Transaction> = createBridges(executeWithSigner(createSigner), network);

const resolver: Resolver = createResolver(createSigner, algorandResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, resolver };
