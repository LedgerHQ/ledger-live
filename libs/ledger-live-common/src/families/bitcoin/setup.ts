// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { firstValueFrom, from } from "rxjs";
import Transport from "@ledgerhq/hw-transport";
import { Bridge } from "@ledgerhq/types-live";
import { CryptoCurrency } from "@ledgerhq/types-cryptoassets";
import Btc from "@ledgerhq/hw-app-btc";
import { createBridges } from "@ledgerhq/coin-bitcoin/bridge/js";
import type { SignerContext } from "@ledgerhq/coin-bitcoin/signer";
import makeCliTools from "@ledgerhq/coin-bitcoin/cli-transaction";
import bitcoinResolver from "@ledgerhq/coin-bitcoin/hw-getAddress";
import { signMessage } from "@ledgerhq/coin-bitcoin/hw-signMessage";
import { Transaction } from "@ledgerhq/coin-bitcoin/types";
import { GetAddressOptions, Resolver } from "../../hw/getAddress/types";
import { withDevice } from "../../hw/deviceAccess";
import { startSpan } from "../../performance";
import { GetAddressFn } from "@ledgerhq/coin-framework/bridge/getAddressWrapper";
import { getCurrencyConfiguration } from "../../config";
import { BitcoinConfigInfo } from "@ledgerhq/coin-bitcoin/lib/config";
import { SignMessage } from "../../hw/signMessage/types";

const createSigner = (transport: Transport, currency: CryptoCurrency) => {
  return new Btc({ transport, currency: currency.id });
};

const signerContext: SignerContext = <T>(
  deviceId: string,
  crypto: CryptoCurrency,
  fn: (signer: Btc) => Promise<T>,
): Promise<T> =>
  firstValueFrom(
    withDevice(deviceId)((transport: Transport) => from(fn(createSigner(transport, crypto)))),
  );

const getCurrencyConfig = (currency: CryptoCurrency) => {
  return { info: getCurrencyConfiguration<BitcoinConfigInfo>(currency) };
};

const perfLogger = {
  startSpan,
};

const bridge: Bridge<Transaction> = createBridges(signerContext, perfLogger, getCurrencyConfig);

export function createMessageSigner(): SignMessage {
  return (transport, account, messageData) => {
    const signerContext: SignerContext = (_, crypto, fn) => fn(createSigner(transport, crypto));
    return signMessage(signerContext)("", account, messageData);
  };
}

const messageSigner = {
  signMessage: createMessageSigner(),
};

const resolver: Resolver = (
  transport: Transport,
  addressOpt: GetAddressOptions,
): ReturnType<GetAddressFn> => {
  const signerContext: SignerContext = (_, crypto, fn) => fn(createSigner(transport, crypto));
  return bitcoinResolver(signerContext)("", addressOpt);
};

const cliTools = makeCliTools();

export { bridge, cliTools, resolver, messageSigner, signerContext };
