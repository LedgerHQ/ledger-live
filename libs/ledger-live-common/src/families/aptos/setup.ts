// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import {
  AptosAccount,
  TransactionStatus,
  createBridges,
  type Transaction,
} from "@ledgerhq/coin-aptos";
import Transport from "@ledgerhq/hw-transport";
import Aptos from "@ledgerhq/hw-app-aptos";
import type { Bridge } from "@ledgerhq/types-live";
import aptosResolver from "@ledgerhq/coin-aptos/signer/index";
import makeCliTools from "@ledgerhq/coin-aptos/test/cli";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import { Resolver } from "../../hw/getAddress/types";

const createSigner: CreateSigner<Aptos> = (transport: Transport) => {
  return new Aptos(transport);
};

const bridge: Bridge<Transaction, AptosAccount, TransactionStatus> = createBridges(
  executeWithSigner(createSigner),
);

const resolver: Resolver = createResolver(createSigner, aptosResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, resolver };
