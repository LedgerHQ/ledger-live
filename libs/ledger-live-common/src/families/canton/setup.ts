// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { createBridges } from "@ledgerhq/coin-canton/bridge/index";
import { getEnv } from "@ledgerhq/live-env";
import Transport from "@ledgerhq/hw-transport";
import { CantonSigner } from "@ledgerhq/coin-canton";
import cantonResolver from "@ledgerhq/coin-canton/signer";
import type { Account, Bridge } from "@ledgerhq/types-live";
import makeCliTools from "@ledgerhq/coin-canton/test/cli";
import { CantonCoinConfig } from "@ledgerhq/coin-canton/config";
import { TransactionStatus, Transaction } from "@ledgerhq/coin-canton/types";
import { LegacySignerCanton } from "@ledgerhq/live-signer-canton";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import cantonBridgeMock from "./bridge/mock";
import { Resolver } from "../../hw/getAddress/types";
import { getCurrencyConfiguration } from "../../config";
import { getCryptoCurrencyById } from "../../currencies";

const createSigner: CreateSigner<CantonSigner> = (transport: Transport) => {
  return new LegacySignerCanton(transport);
};

const getCurrencyConfig = () => {
  // Use devnet for development/testing
  const currencyId = getEnv("MOCK") ? "canton_network_devnet" : "canton_network_devnet";
  return getCurrencyConfiguration<CantonCoinConfig>(getCryptoCurrencyById(currencyId));
};

const bridge: Bridge<Transaction, Account, TransactionStatus> = getEnv("MOCK")
  ? cantonBridgeMock
  : createBridges(executeWithSigner(createSigner), getCurrencyConfig);

const resolver: Resolver = createResolver(createSigner, cantonResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, resolver };
