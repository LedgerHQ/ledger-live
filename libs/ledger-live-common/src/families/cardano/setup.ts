// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import Ada, { Networks, AddressType } from "@cardano-foundation/ledgerjs-hw-app-cardano";
import { str_to_path } from "@cardano-foundation/ledgerjs-hw-app-cardano/dist/utils/address";
import Transport from "@ledgerhq/hw-transport";
import type { Bridge } from "@ledgerhq/types-live";
import { createBridges } from "@ledgerhq/coin-cardano/bridge";
import makeCliTools from "@ledgerhq/coin-cardano/cli-transaction";
import cardanoResolver from "@ledgerhq/coin-cardano/hw-getAddress";
import type {
  CardanoAccount,
  CardanoLikeNetworkParameters,
  Transaction,
  TransactionStatus,
} from "@ledgerhq/coin-cardano/types";
import type {
  CardanoAddress,
  CardanoExtendedPublicKey,
  CardanoSignRequest,
  CardanoSignature,
  CardanoSigner,
  GetAddressRequest,
} from "@ledgerhq/coin-cardano/signer";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import type { Resolver } from "../../hw/getAddress/types";
import signerSerializer from "./signerSerializer";
import { getCurrencyConfiguration } from "../../config";
import { CardanoCoinConfig } from "@ledgerhq/coin-cardano/config";
import { getCryptoCurrencyById } from "@ledgerhq/cryptoassets/currencies";

function findNetwork(networkParams: CardanoLikeNetworkParameters) {
  return networkParams.networkId === Networks.Mainnet.networkId
    ? Networks.Mainnet
    : Networks.Testnet;
}

const createSigner: CreateSigner<CardanoSigner> = (transport: Transport) => {
  const ada = new Ada(transport);
  return {
    getAddress: async ({
      path,
      stakingPathString,
      networkParams,
      verify,
    }: GetAddressRequest): Promise<CardanoAddress> => {
      const network = findNetwork(networkParams);

      const addr = await ada.deriveAddress({
        network,
        address: {
          type: AddressType.BASE_PAYMENT_KEY_STAKE_KEY,
          params: {
            spendingPath: str_to_path(path),
            stakingPath: str_to_path(stakingPathString),
          },
        },
      });
      if (verify) {
        await ada.showAddress({
          network,
          address: {
            type: AddressType.BASE_PAYMENT_KEY_STAKE_KEY,
            params: {
              spendingPath: str_to_path(path),
              stakingPath: str_to_path(stakingPathString),
            },
          },
        });
      }

      return addr;
    },
    getPublicKey: async (accountPath: string): Promise<CardanoExtendedPublicKey> => {
      return ada.getExtendedPublicKey({
        path: str_to_path(accountPath),
      });
    },
    sign: async ({ transaction, networkParams }: CardanoSignRequest): Promise<CardanoSignature> => {
      const network = findNetwork(networkParams);
      const trxOptions = signerSerializer(network, transaction);

      return ada.signTransaction(trxOptions);
    },
  };
};

const getCurrencyConfig = () => {
  return getCurrencyConfiguration<CardanoCoinConfig>(getCryptoCurrencyById("cardano"));
};

const bridge: Bridge<Transaction, CardanoAccount, TransactionStatus> = createBridges(
  executeWithSigner(createSigner),
  getCurrencyConfig,
);

const resolver: Resolver = createResolver(createSigner, cardanoResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, resolver };
