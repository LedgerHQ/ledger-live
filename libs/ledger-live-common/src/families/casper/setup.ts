// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { createBridges } from "@ledgerhq/coin-casper/bridge";
import Transport from "@ledgerhq/hw-transport";
import Casper from "@zondax/ledger-casper";
import casperResolver from "@ledgerhq/coin-casper/signer";
import { signMessage } from "@ledgerhq/coin-casper/hw-signMessage";
import type { Account, Bridge } from "@ledgerhq/types-live";
import makeCliTools from "@ledgerhq/coin-casper/test/cli";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import { Resolver } from "../../hw/getAddress/types";
import { TransactionStatus, Transaction } from "@ledgerhq/coin-casper/types";
import { CasperGetAddrResponse, CasperSignature, CasperSigner } from "./types";
import { getCurrencyConfiguration } from "../../config";
import { getCryptoCurrencyById } from "@ledgerhq/cryptoassets/currencies";
import { getPath, isError } from "./common";
import { CasperCoinConfig } from "@ledgerhq/coin-casper/config";

const createSigner: CreateSigner<CasperSigner> = (transport: Transport) => {
  const casper = new Casper(transport);
  return {
    showAddressAndPubKey: async (path: string): Promise<CasperGetAddrResponse> => {
      const r = await casper.showAddressAndPubKey(getPath(path));
      isError(r);

      return r;
    },
    getAddressAndPubKey: async (path: string): Promise<CasperGetAddrResponse> => {
      const r = await casper.getAddressAndPubKey(getPath(path));
      isError(r);

      return r;
    },
    sign: async (path: string, message: Buffer): Promise<CasperSignature> => {
      const r = await casper.sign(getPath(path), message);
      isError(r);

      return r;
    },
  };
};

const getCoinConfig: CasperCoinConfig = () =>
  getCurrencyConfiguration<ReturnType<CasperCoinConfig>>(getCryptoCurrencyById("casper"));

const bridge: Bridge<Transaction, Account, TransactionStatus> = createBridges(
  executeWithSigner(createSigner),
  getCoinConfig,
);

const messageSigner = {
  signMessage,
};

const resolver: Resolver = createResolver(createSigner, casperResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, messageSigner, resolver };
