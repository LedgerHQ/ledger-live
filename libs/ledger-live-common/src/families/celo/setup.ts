// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { createBridges } from "@ledgerhq/coin-celo/bridge";
import makeCliTools from "@ledgerhq/coin-celo/cli-transaction";
import CeloResolver from "@ledgerhq/coin-celo/hw-getAddress";
import Transport from "@ledgerhq/hw-transport";
import type { Bridge } from "@ledgerhq/types-live";
import Celo from "@ledgerhq/hw-app-celo";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import type { Resolver } from "../../hw/getAddress/types";
import { Transaction, CeloAccount } from "./types";
import { CeloSigner } from "@ledgerhq/coin-celo/signer";

const createSigner: CreateSigner<CeloSigner> = (transport: Transport) => new Celo(transport);

const bridge: Bridge<Transaction, CeloAccount> = createBridges(executeWithSigner(createSigner));

const resolver: Resolver = createResolver(createSigner, CeloResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, resolver };
