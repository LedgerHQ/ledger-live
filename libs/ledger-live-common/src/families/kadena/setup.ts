// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { createBridges } from "@ledgerhq/coin-kadena/bridge/js";
import makeCliTools from "@ledgerhq/coin-kadena/cli-transaction";
import kadenaResolver from "@ledgerhq/coin-kadena/hw-getAddress";
import { Transaction } from "@ledgerhq/coin-kadena/types";
import Kadena from "@zondax/hw-app-kda";
import Transport from "@ledgerhq/hw-transport";
import type { Bridge } from "@ledgerhq/types-live";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import type { Resolver } from "../../hw/getAddress/types";
import { KadenaSigner } from "@ledgerhq/coin-kadena/lib/signer";

const createSigner: CreateSigner<KadenaSigner> = (transport: Transport) => {
  return new Kadena(transport as any);
};

const bridge: Bridge<Transaction> = createBridges(executeWithSigner(createSigner));

const resolver: Resolver = createResolver(createSigner, kadenaResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, resolver };
