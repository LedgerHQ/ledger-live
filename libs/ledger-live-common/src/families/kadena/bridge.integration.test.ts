import { dataset } from "@ledgerhq/coin-kadena/__tests__/integration/bridge.integration.test";
import { testBridge } from "../../__tests__/test-helpers/bridge";
import "../../__tests__/test-helpers/setup";

testBridge(dataset);
