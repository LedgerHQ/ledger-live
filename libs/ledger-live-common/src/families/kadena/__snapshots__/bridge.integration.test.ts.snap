// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kadena currency bridge scanAccounts kadena seed 1 1`] = `
[
  {
    "balance": "17998528000000",
    "currencyId": "kadena",
    "derivationMode": "",
    "freshAddress": "k:15daab7d9ba9f8a465ffc4bfb33e68ca5e9f51ef0bab387284963129fe04ec3e",
    "freshAddressPath": "44'/626'/0'/0/0",
    "id": "js:2:kadena:15daab7d9ba9f8a465ffc4bfb33e68ca5e9f51ef0bab387284963129fe04ec3e:",
    "index": 0,
    "operationsCount": 3,
    "pendingOperations": [],
    "seedIdentifier": "15daab7d9ba9f8a465ffc4bfb33e68ca5e9f51ef0bab387284963129fe04ec3e",
    "spendableBalance": "17998528000000",
    "swapHistory": [],
    "syncHash": undefined,
    "used": true,
    "xpub": "15daab7d9ba9f8a465ffc4bfb33e68ca5e9f51ef0bab387284963129fe04ec3e",
  },
  {
    "balance": "0",
    "currencyId": "kadena",
    "derivationMode": "",
    "freshAddress": "k:ced7b6fcaf2421b294f431965374d43e3b97352131609162d6c78325ad609015",
    "freshAddressPath": "44'/626'/1'/0/0",
    "id": "js:2:kadena:ced7b6fcaf2421b294f431965374d43e3b97352131609162d6c78325ad609015:",
    "index": 1,
    "operationsCount": 0,
    "pendingOperations": [],
    "seedIdentifier": "15daab7d9ba9f8a465ffc4bfb33e68ca5e9f51ef0bab387284963129fe04ec3e",
    "spendableBalance": "0",
    "swapHistory": [],
    "syncHash": undefined,
    "used": false,
    "xpub": "ced7b6fcaf2421b294f431965374d43e3b97352131609162d6c78325ad609015",
  },
]
`;

exports[`kadena currency bridge scanAccounts kadena seed 1 2`] = `
[
  [
    {
      "accountId": "js:2:kadena:15daab7d9ba9f8a465ffc4bfb33e68ca5e9f51ef0bab387284963129fe04ec3e:",
      "blockHash": "O5FF3qB3_wmrdTxucueqec4XSK4NvkiGYVZV2BrkTuU",
      "blockHeight": 4270944,
      "extra": {
        "receiverChainId": 0,
        "senderChainId": 0,
      },
      "fee": "20000000000000",
      "hash": "8qgaZP0nWHd-O7n2fyDF6LVbQY89kDTO1mO9xSN3fdo",
      "id": "js:2:kadena:15daab7d9ba9f8a465ffc4bfb33e68ca5e9f51ef0bab387284963129fe04ec3e:-8qgaZP0nWHd-O7n2fyDF6LVbQY89kDTO1mO9xSN3fdo-IN",
      "recipients": [
        "k:15daab7d9ba9f8a465ffc4bfb33e68ca5e9f51ef0bab387284963129fe04ec3e",
      ],
      "senders": [
        "c:Ecwy85aCW3eogZUnIQxknH8tG8uXHM5QiC__jeI0nWA",
      ],
      "type": "IN",
      "value": "20000000000000",
    },
    {
      "accountId": "js:2:kadena:15daab7d9ba9f8a465ffc4bfb33e68ca5e9f51ef0bab387284963129fe04ec3e:",
      "blockHash": "mOMiT32h-td1YgpXcw51or6iwo8IZrBeKdTUTQSnle0",
      "blockHeight": 4290505,
      "extra": {
        "receiverChainId": 0,
        "senderChainId": 0,
      },
      "fee": "736000000",
      "hash": "AdiO7TIUc72AVQ9q8ByKgc8A0PNrwULgeqBc6Z5bfpk",
      "id": "js:2:kadena:15daab7d9ba9f8a465ffc4bfb33e68ca5e9f51ef0bab387284963129fe04ec3e:-AdiO7TIUc72AVQ9q8ByKgc8A0PNrwULgeqBc6Z5bfpk-OUT",
      "recipients": [
        "k:77b021744ab3c003e8e4d0f38a598f0e39fe9a7fe61360754dc7321b112ab375",
      ],
      "senders": [
        "k:15daab7d9ba9f8a465ffc4bfb33e68ca5e9f51ef0bab387284963129fe04ec3e",
      ],
      "type": "OUT",
      "value": "1000000000000",
    },
    {
      "accountId": "js:2:kadena:15daab7d9ba9f8a465ffc4bfb33e68ca5e9f51ef0bab387284963129fe04ec3e:",
      "blockHash": "1OrVeC9HWVpWFVwTzHT8_nJoDiHYV5EIHeqwm3WoFzc",
      "blockHeight": 4290491,
      "extra": {
        "receiverChainId": 0,
        "senderChainId": 0,
      },
      "fee": "736000000",
      "hash": "ukMO2OihPUy6G4ukJjAUcYy-tnf3WgdKIMbsFu8zrvk",
      "id": "js:2:kadena:15daab7d9ba9f8a465ffc4bfb33e68ca5e9f51ef0bab387284963129fe04ec3e:-ukMO2OihPUy6G4ukJjAUcYy-tnf3WgdKIMbsFu8zrvk-OUT",
      "recipients": [
        "k:77b021744ab3c003e8e4d0f38a598f0e39fe9a7fe61360754dc7321b112ab375",
      ],
      "senders": [
        "k:15daab7d9ba9f8a465ffc4bfb33e68ca5e9f51ef0bab387284963129fe04ec3e",
      ],
      "type": "OUT",
      "value": "1000000000000",
    },
  ],
  [],
]
`;
